source/Messages.cpp

```diff
@@ -27,34 +27,31 @@ namespace {
 
 	mutex incomingMutex;
 
-	vector<pair<string, Messages::Importance>> incoming;
+	vector<pair<string, const Message::Category *>> incoming;
 	vector<Messages::Entry> recent;
-	deque<pair<string, Messages::Importance>> logged;
+	deque<pair<string, const Message::Category *>> logged;
 }
 
 
 
-// Add a message to the list along with its level of importance
-// When forced, the message is forcibly added to the log, but not to the list.
-void Messages::Add(const string &message, Importance importance, bool force)
+// Add a message to the list along with its level of importance.
+void Messages::Add(const Message &message)
 {
-	lock_guard<mutex> lock(incomingMutex);
-	incoming.emplace_back(message, importance);
-	AddLog(message, importance, force);
-}
-
-
-
-// Add a message to the log. For messages meant to be shown
-// also on the main panel, use Add instead.
-void Messages::AddLog(const string &message, Importance importance, bool force)
-{
-	if(force || logged.empty() || message != logged.front().first)
+	const Message::Category *category = message.GetCategory();
+	if(!category)
+		return;
+	string text = message.Text();
+	if(!category->LogDeduplication() || logged.empty() || text != logged.front().first)
 	{
-		logged.emplace_front(message, importance);
+		logged.emplace_front(text, category);
 		if(logged.size() > MAX_LOG)
 			logged.pop_back();
 	}
+
+	if(category->LogOnly())
+		return;
+	lock_guard<mutex> lock(incomingMutex);
```

Comment:
Should the mutex be locked before the emplace_front of logged?

Reply:
We don't have any mutex for logged, as it's only for the message log panel, and no messages are added while the panel is open.

Reply:
Right, the original `AddLog` message just didn't have any mutex locking.